### PR TITLE
 ethvm: Do not chenge VM selection depending on tracing

### PR DESCRIPTION
**Pull request by chfast**
Mostly: do not force Interpreter if VM tracing is required. Also cleanups some parts of ethvm.

Fixes https://github.com/ethereum/cpp-ethereum/issues/4899.
Fixes https://github.com/ethereum/cpp-ethereum/issues/4795.

### CODE REVIEW

File: ethvm/main.cpp [hunk 1]
```diff
@@ -345,23 +350,28 @@ int main(int argc, char** argv)
...
         cout << "Expensive operations:\n";
-        for (auto const& c: {Instruction::SSTORE, Instruction::SLOAD, Instruction::CALL, Instruction::CREATE, Instruction::CALLCODE, Instruction::DELEGATECALL, Instruction::MSTORE8, Instruction::MSTORE, Instruction::MLOAD, Instruction::SHA3})
-            if (!!counts[(byte)c].first)
-                cout << "  " << instructionInfo(c).name << " x " << counts[(byte)c].first << " (" << counts[(byte)c].second << " gas)\n";
+        for (auto const& c : {Instruction::SSTORE, Instruction::SLOAD, Instruction::CALL,
+                 Instruction::CREATE, Instruction::CALLCODE, Instruction::DELEGATECALL,
+                 Instruction::MSTORE8, Instruction::MSTORE, Instruction::MLOAD, Instruction::SHA3})
+            if (counts[(byte)c].first != 0)
+                cout << "  " << instructionInfo(c).name << " x " << counts[(byte)c].first << " ("
+                     << counts[(byte)c].second << " gas)\n";
```
Comment (axic) [inline ethvm/main.cpp]: `static_cast` for `c`, and perhaps a local variable instead of 3 casts?

**Comment by axic** [comment]
Should a warning be shown if `tracing` is selected but the VM ignores it?

File: ethvm/main.cpp [hunk 2]
```diff
@@ -316,32 +315,30 @@ int main(int argc, char** argv)
...
 
+    OnOpFunc onOp;
+    if (mode == Mode::Statistics)
+    {
+        onOp = [&](uint64_t, uint64_t, Instruction inst, bigint m, bigint gasCost, bigint,
+                   VMFace const*, ExtVMFace const*) {
+            counts[(byte)inst].first++;
+            counts[(byte)inst].second += gasCost;
+            total++;
+            if (m > memTotal)
```
Comment (axic) [inline ethvm/main.cpp]: This line was probably a bugfix worth a separate commit.

**Comment by chfast** [comment]
> Should a warning be shown if tracing is selected but the VM ignores it?

This is hard to do at this point because this is constantly changing. E.g. Hera supports tracing but not through EVM-C.

**Comment by axic** [comment]
> E.g. Hera supports tracing but not through EVM-C.

It actually doesn't support tracing, only pseudo tracing for evm2wasm compiled EVM code :)

Perhaps the future evm-c tracing API should be more flexible and/or have a kind/flavour option.

**Comment by chfast** [comment]
Long term: we can report back if a VM supports tracing.
Short term: we can add some ifs and warnings, but I wouldn't do it until someone complains.

**Comment by chfast** [comment]
Thanks for checking...

**Pull request merged**